### ISSUE TITLE
Why don't the Reflection.newProxy add Class Proxy like cglib?

### ISSUE THREAD

**Issue by cao19920425** [google/guava#2982]
I think the add Class proxy will be better.
Why don't the add Class Proxy like cglib?

**Comment by cpovirk** [comment]
A few factors:

- We've mostly avoided depending on other libraries.
- cglib doesn't work on Android.
- cglib has [an API similar to `Reflection.newProxy`](http://cglib.sourceforge.net/apidocs/net/sf/cglib/Enhancer.html#create(java.lang.Class,%20net.sf.cglib.Callback)) already.

**Comment by cao19920425** [comment]
Thank you for your answer.
Your means that now all Methods in guava can work on Android?

**Comment by cpovirk** [comment]
We expect for anything in the `-android` variant of Guava to work on Android. (We've occasionally found bugs in very strange edge cases, but we don't expect users to encounter them in practice.) That said, we don't recommend using, e.g., reflection, including APIs like `eventbus` that use reflection, because of poor performance on Android.

**Comment by cao19920425** [comment]
um...
But reflection is the important feature of Java.
If the situation that APIs of reflection poor performance on Android can't be improved.
May divide guava into generic and unique modules be better?

**Comment by cpovirk** [comment]
If we could go back in time, we might well not have included the reflection utilities in our `-android` variant. But previously we didn't have a separate variant, so any classes we included needed to be available on Android. Now that we've released them there, we don't want to pull them back and break anyone who uses them there.

(There's another complication, which is that we recommend using the "Android" variant if you're stuck on Java 7. Reflection could be useful there, too.)
